### DISCUSSION TITLE
Typing module.hot for TypeScript

### DISCUSSION THREAD

**Discussion by edemaine** [meteor/meteor#11836] [Feature Requests]
`module.hot` from [hot-module-replacement](https://docs.meteor.com/packages/hot-module-replacement.html) is undeclared from a TypeScript perspective.  I could manually declare it as follows:

```ts
interface MeteorModule extends NodeModule {
  hot?: {
    accept(): void;
    reject(): void;
    dispose(callback: (data: any) => void): void;
    data: any;
    onRequire(callbacks: {
      before(requiredModule: MeteorModule, parentId: string): any;
      after(requiredModule: MeteorModule, data: any): void;
    }): void;
  }
};
declare var module: MeteorModule;
```

(I think I got everything right, but it's worth double checking.)

I'm unfamiliar with contributing to DefinitelyTyped, and perhaps the PR should come from a Meteor dev?

Modeling after [`types/node/globals.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/globals.d.ts), it looks like we need a new file in [`types/meteor`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/meteor) called `global.d.ts` with the declaration above, and then users' `tsconfig.json` will need to change from

```json
    "types": ["node"],
```
to
```json
    "types": ["node", "meteor"],
```
Is this correct?

@zodern

**Comment by radekmie** [comment]
Typing `module.hot` is a good idea! We've also started using it here and there and for now we simply relied on some hand-crafted types, similar to yours.

A few notes though:
1. As the [docs](https://docs.meteor.com/packages/hot-module-replacement.html#module-hot-decline) and [code](https://github.com/meteor/meteor/blob/3ae22ca82096e500c4c29f8eba0799e55d428bd0/packages/hot-module-replacement/hot-api.js#L44-L59) say - it's `decline`, not `reject`.
2. Use `unknown` instead of `any`. It's safer and if needed, one can cast it anyhow.
3. According to the [`Module.prototype.require`](https://github.com/meteor/meteor/blob/dae7af832d08a8b19384ac19aa5a5a9b6b005e55/packages/modules-runtime-hot/installer.js#L229-L265) implementation, both `before` and `after` properties of the `callbacks` argument of `onRequire` are be optional.
4. I'd actually suggest `hot-module-replacement.d.ts` to match the package name, just like the other files.

With these changes, you can file a PR to DefinitelyTyped and any of the package owners (including myself) will be able to review it. Just remember to add tests as well, so we can merge it without DefinitelyTyped maintainers!

Also, in most cases, you don't need the `types` property in `tsconfig.json` at all. By default, all `@types` packages are included.

---

It looks like [this `return`](https://github.com/meteor/meteor/blob/3ae22ca82096e500c4c29f8eba0799e55d428bd0/packages/hot-module-replacement/hot-api.js#L81) is not needed, as [`_onRequire` doesn't return anything](https://github.com/meteor/meteor/blob/dae7af832d08a8b19384ac19aa5a5a9b6b005e55/packages/modules-runtime-hot/installer.js#L215-L217).

**Reply by edemaine** [reply]
Thanks for all the improvements and tips!  I ended up having to tweak the definitions a fair amount to avoid conflicts with Node, and to allow for the `data` field to be typeable by the user.  (I also realized that the `data` argument in `after` is different from the other uses of `data: it's the return value of `before`.)

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58014

**Reply by edemaine** [reply]
FYI, this is now done and released!
